gensim/models/fasttext.py

```diff
@@ -9,77 +9,83 @@
-                       model.wv.vocab[w].sample_int > model.random.rand() * 2**32]
-        for pos, word in enumerate(word_vocabs):
-            reduced_window = model.random.randint(model.window)
-            start = max(0, pos - model.window + reduced_window)
-            window_pos = enumerate(word_vocabs[start:(pos + model.window + 1 - reduced_window)], start)
-            word2_indices = [word2.index for pos2, word2 in window_pos if (word2 is not None and pos2 != pos)]
-
-            word2_subwords = []
-            vocab_subwords_indices = []
-            ngrams_subwords_indices = []
-
-            for index in word2_indices:
-                vocab_subwords_indices += [index]
-                word2_subwords += model.wv.ngrams_word[model.wv.index2word[index]]
-
-            for subword in word2_subwords:
-                ngrams_subwords_indices.append(model.wv.ngrams[subword])
-
-            l1_vocab = np_sum(model.wv.syn0_vocab[vocab_subwords_indices], axis=0)  # 1 x vector_size
-            l1_ngrams = np_sum(model.wv.syn0_ngrams[ngrams_subwords_indices], axis=0)  # 1 x vector_size
-
-            l1 = np_sum([l1_vocab, l1_ngrams], axis=0)
-            subwords_indices = [vocab_subwords_indices] + [ngrams_subwords_indices]
-            if (subwords_indices[0] or subwords_indices[1]) and model.cbow_mean:
-                l1 /= (len(subwords_indices[0]) + len(subwords_indices[1]))
-
-            # train on the sliding window for target word
-            train_cbow_pair(model, word, subwords_indices, l1, alpha, is_ft=True)
-        result += len(word_vocabs)
-    return result
-
-
-def train_batch_sg(model, sentences, alpha, work=None):
-    result = 0
-    for sentence in sentences:
-        word_vocabs = [model.wv.vocab[w] for w in sentence if w in model.wv.vocab and
-                       model.wv.vocab[w].sample_int > model.random.rand() * 2**32]
-        for pos, word in enumerate(word_vocabs):
-            reduced_window = model.random.randint(model.window)  # `b` in the original word2vec code
-            # now go over all words from the (reduced) window, predicting each one in turn
-            start = max(0, pos - model.window + reduced_window)
-
-            subwords_indices = [word.index]
-            word2_subwords = model.wv.ngrams_word[model.wv.index2word[word.index]]
-
-            for subword in word2_subwords:
-                subwords_indices.append(model.wv.ngrams[subword])
-
-            for pos2, word2 in enumerate(word_vocabs[start:(pos + model.window + 1 - reduced_window)], start):
-                if pos2 != pos:  # don't train on the `word` itself
-                    train_sg_pair(model, model.wv.index2word[word2.index], subwords_indices, alpha, is_ft=True)
-
-        result += len(word_vocabs)
-    return result
+try:
+    from gensim.models.fasttext_inner import train_batch_sg, train_batch_cbow
+    from gensim.models.fasttext_inner import FAST_VERSION, MAX_WORDS_IN_BATCH
+
+except ImportError:
```

Comment:
The problem with this solution is that it makes it impossible to test the Python native implementation. There's a number of ways you could fix this. The first thing that comes into my mind is this:

- add another module `fasttext_native` or `fasttext_legacy` and define the native version of `train_batch_sg, train_batch_cbow, FAST_VERSION, MAX_WORDS_IN_BATCH` there
- import as follows:

```python
try:
    from gensim.models.fasttext_inner import train_batch_sg, train_batch_cbow, FAST_VERSION, MAX_WORDS_IN_BATCH
except ImportError:
    # log warning
    from gensim.models.fasttext_native import train_batch_sg, train_batch_cbow, FAST_VERSION, MAX_WORDS_IN_BATCH
```

- add two new params to `FastText.__init__()` for `train_batch_sg` and `train_batch_cbow` functions. The default values would be the imported functions.

To test the native/legacy implementation, you would then be able to initialize `FastText` as follows:

```python
from gensim.models.fasttext import FastText
from gensim.models.fasttext_native import MAX_WORDS_IN_BATCH, train_batch_sg, train_batch_cbow

ft = FastText(..., batch_words=MAX_WORDS_IN_BATCH, train_batch_sg=train_batch_sg, train_batch_cbow=train_batch_cbow)
# test ft
```

I don't insist on this particular solution. There may be more elegant ways. My main concern is that the native implementation should also be testable and covered by unit tests (just parametrize existing unit tests to test both the native and Cython implementation).

Reply:
Native Python implementations will be removed in the next refactor. We're not going to support them any more, now that @menshikh-iv can build good wheels.

Reply:
I hope that this will be true @piskvorky, need to distribute our wheels several releases in row first, and if all will be OK - we can remove python parts (but this process is not fast, I suggest not rushing with this)

Reply:
@janpom I think this code looks exactly repeat logic of https://github.com/RaRe-Technologies/gensim/blob/develop/gensim/models/word2vec.py#L137.

Reply:
What I mean is there's no point creating new constructions for testing both versions, when we know we only want one version.